File: 322-coin-change/coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    let l = coins.length
    const dp = [...new Array(amount+1)].fill(Infinity)
    dp[0] = 0

    for (let c of coins) { // usually the outter loop
        for (let r = 1; r <= amount; r++) { // usually the inner loop
            if ( r - c >= 0) {
                dp[r] = Math.min(dp[r], dp[r - c] + 1)
            }
        }
    }
    return dp[amount] === Infinity ? -1 : dp[amount]
};

File: 322-coin-change/coin-change.java
class Solution {
    // Bottom-Up Dynamic Programming.
    // Time and Space: O(n * m)...n for coins and m for amount
    // Space: O(m)...m for amount
    public int coinChange(int[] coins, int amount) {
        int dp[] = new int[amount+1];
        Arrays.fill(dp, amount+1);
        dp[0] = 0; // base case
        for (int c : coins) {
            for (int a = 1; a <= amount; a++) {
                if (a - c >= 0) {
                    dp[a] = Math.min(dp[a], dp[a-c] + 1);
                }
            }
         }
        return dp[amount] == amount+1 ? -1 : dp[amount];
    }
}